src/main/java/duke/task/TaskManager.java

```diff
@@ -0,0 +1,252 @@
+    private boolean hasChanged = false;
+
+
+    /**
+     * Creates a manager.
+     */
+    public TaskManager() {
+        this.items = new ArrayList<>();
+    }
+
+    /**
+     * Adds task to the current manager list.
+     *
+     * @param item A task to add.
+     */
+    public void addItem(Task item) {
+        items.add(item);
+    }
+
+    /**
+     * Adds a Task based on the possible Actions and instruction to the current manage
+     *
+     * @param options     A valid Action to perform.
+     * @param instruction A string value of the commands.
+     * @return An ArrayList of string of the values to output to the Ui.
+     * @throws DukeException Invalid processing of the items.
+     */
+    public ArrayList<String> addTask(Actions options, String instruction) throws DukeException {
+        Task item;
+        String description;
+        String by;
+        String from;
+        Pattern deadlineFormat = Pattern.compile("(?<description>.+)\\s?((?i)/by)(?<by>.+)");
+        Pattern eventFormat = Pattern.compile("(?<description>.+)\\s?((?i)/from)(?<from>.+)((?i)/to)(?<by>.+)");
+        switch (options) {
+        case TODO:
+            if (instruction.isBlank()) {
+                throw new DukeException("description");
+            }
+            item = new Todo(instruction);
+            break;
+        case DEADLINE:
+            Matcher deadlineMatch = deadlineFormat.matcher(instruction);
+            if (!deadlineMatch.find()) {
+                throw new DukeException("GIGABOOOM");
+            }
+            description = deadlineMatch.group("description");
+            by = deadlineMatch.group("by").trim();
+            if (description.isBlank()) {
+                throw new DukeException("description");
+            } else if (by.isBlank()) {
+                throw new DukeException("by");
+            }
+            Optional<LocalDate> testDate = DateHandler.checkDate(by);
+            item = testDate.map(localDate -> new Deadline(description, LocalDateTime.of(localDate, DateHandler
+                    .checkTime(by)
+                    .orElse(LocalTime.of(0, 0))))).orElseGet(() -> new Deadline(description, by));
+            break;
+        case EVENT:
+            Matcher eventMatch = eventFormat.matcher(instruction);
+            if (!eventMatch.find()) {
+                System.out.println("reached here");
+                throw new DukeException("GIGABOOOM");
+            }
+            description = eventMatch.group("description");
+            by = eventMatch.group("by").trim();
+            from = eventMatch.group("from").trim();
+            if (from.isBlank()) {
+                throw new DukeException("from");
+            } else if (by.isBlank()) {
+                throw new DukeException("by");
+            } else if (description.isBlank()) {
+                throw new DukeException("description");
+            }
+            Optional<LocalDate> testByDate = DateHandler.checkDate(by);
```

Comment:
i would have never thought of using Optional and FlatMap and i think it is quite cool that you managed to incorporate it into your IP! 💯